cmd/block-generator/main.go

```diff
@@ -0,0 +1,114 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"math/rand"
+	"net/http"
+	"os"
+	"strings"
+
+	"github.com/spf13/viper"
+
+	"github.com/algorand/indexer/cmd/block-generator/generator"
+	"github.com/algorand/indexer/util"
+)
+
+var configFile string
+var port uint
+var gen generator.Generator
+
+const configFileName = "block_generator_config"
+
+func init() {
+	rand.Seed(12345)
+	flag.StringVar(&configFile, "config", "", fmt.Sprintf("Override default config file from '%s'.", configFileName))
+	flag.UintVar(&port, "port", 4010, "Port to start the server at.")
+
+	viper.SetConfigName(configFileName)
+	viper.SetConfigType("yaml")
+	viper.AddConfigPath(".")
+}
+
+func initializeConfigFile() error {
+	if len(configFile) > 0 {
+		f, err := os.Open(configFile)
+		if err != nil {
+			return err
+		}
+
+		return viper.ReadConfig(f)
+	}
+
+	return viper.ReadInConfig()
+}
+
+func main() {
+	flag.Parse()
+
+	util.MaybeFail(initializeConfigFile(), "problem loading config file. Use '-config' or create a config file.")
+
+	// Pass everything from the configuration into the generator.
+	gen = generator.MakeGenerator(generator.GenerationConfig{
+		TxnPerBlock:                  15000,
+		NewAccountFrequency:          100,
+		Protocol:                     "future",
+		NumGenesisAccounts:           10,
+		GenesisAccountInitialBalance: 1000000000,
+		GenesisID:                    "blockgen-test",
+		GenesisHash:                  [32]byte{},
+	})
+
+	http.HandleFunc("/", help)
+	http.HandleFunc("/v2/blocks/", handleBlock)
+	http.HandleFunc("/genesis", handleGenesis)
+
+	portStr := fmt.Sprintf(":%d", port)
+	fmt.Printf("Starting server at %s\n", portStr)
+	http.ListenAndServe(portStr, nil)
+}
+
+func help(w http.ResponseWriter, r *http.Request) {
+	fmt.Fprintf(w, "Use /v2/blocks/:blocknum: to get a block.")
+}
+
+func handleGenesis(w http.ResponseWriter, r *http.Request) {
+	gen.WriteGenesis(w)
+}
+
+func handleBlock(w http.ResponseWriter, r *http.Request) {
+	// The generator doesn't actually care about the block...
+	round, err := parseRound(r.URL.Path)
+	if err != nil {
+		fmt.Fprintf(w, err.Error())
+		return
+	}
+
+	gen.WriteBlock(w, round)
+}
+
+const blockQueryPrefix = "/v2/blocks/"
+const blockQueryBlockIdx = len(blockQueryPrefix)
+
+func parseRound(path string) (uint64, error) {
+	if !strings.HasPrefix(path, blockQueryPrefix) {
+		return 0, fmt.Errorf("not a blocks query: %s", path)
+	}
+
+	result := uint64(0)
+	pathlen := len(path)
+
+	if pathlen == blockQueryBlockIdx {
+		return 0, fmt.Errorf("no block in path")
+	}
+
+	for i := blockQueryBlockIdx; i < pathlen; i++ {
+		if path[i] < '0' || path[i] > '9' {
```

Comment:
This can be moved outside of the loop. Also, isn't there some function in go that converts strings to integers?

Reply:
What part can move outside of the loop?

Reply:
The line I commented on - 106. It only does anything if `i = blockQueryBlockIdx`.

Reply:
oh, it's also the exit condition in case there are query parameters like `?pretty`

Reply:
Well, it only exits when `i = blockQueryBlockIdx`. Then it's a bug.

Reply:
can you write some tests?